src/plugins/wizard/server/saved_objects/wizard_migration.ts

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { get, flow } from 'lodash';
+import { SavedObjectMigrationFn } from '../../../../core/server';
+
+const migrateIndexPattern: SavedObjectMigrationFn<any, any> = (doc) => {
+  try {
+    const visualizationStateJSON = get(doc, 'attributes.visualizationState');
+    const visualizationState = JSON.parse(visualizationStateJSON);
+    const indexPatternId = visualizationState.indexPattern;
+    const indexRefName = 'kibanaSavedObjectMeta.searchSourceJSON.index';
+
+    if (indexPatternId && Array.isArray(doc.references)) {
+      const searchSourceIndex = {
+        indexRefName,
+      };
+      const visualizationWithoutIndex = {
+        searchFields: visualizationState.searchFields,
+        activeVisualization: visualizationState.activeVisualization,
+      };
+      doc.attributes.visualizationState = JSON.stringify(visualizationWithoutIndex);
+
+      doc.references.push({
+        name: indexRefName,
+        type: 'index-pattern',
+        id: indexPatternId,
+      });
+
+      return {
+        ...doc,
+        attributes: {
+          ...doc.attributes,
+          kibanaSavedObjectMeta: {
+            searchSourceJSON: JSON.stringify(searchSourceIndex),
+          },
+        },
+      };
+    }
+    return doc;
+  } catch (e) {
+    return doc;
+  }
+};
+
+export const wizardSavedObjectTypeMigrations = {
+  '3.0.0': flow(migrateIndexPattern),
```

Comment:
If we merge this PR into 2.3 release, need to change this to 2.3.0.

3.3.0 here because I am running dashboard 3.3.0 when developing.

Reply:
Since we know that this has to happen for 2.3.0, let's mark it as such. This transformation should happen on only one version

Reply:
Do you know how this works under the hood? All the migrations we have today migrate based on kibanas version which is ahead of OSD. I suspect that this will work as is, but Id like to be sure and hopefully have a plan by the time we get to v6+ when we start having overlaps in versioning.